### PR TITLE
Improve GitHub resolver robustness and efficiency

### LINKED ISSUE #5456: [Bug]: Too many detected issues in github resolver
### Is there an existing issue for the same bug?

- [X] I have checked the existing issues.

### Describe the bug and reproduction steps

There are too many issues detected in the github resolver. For instance, running on this PR: https://github.com/All-Hands-AI/OpenHands/pull/5449

Finds many issues that are not mentioned anywhere in the PR:
```
$ poetry run python openhands/resolver/resolve_issue.py --repo all-hands-ai/openhands --issue-type pr --issue-number 5432
16:15:54 - openhands:INFO: issue_definitions.py:608 - Limiting resolving to issues [5432].
16:15:55 - openhands:WARNING: issue_definitions.py:597 - Failed to fetch issue 6729: 404 Client Error: Not Found for url: https://api.github.com/repos/all-hands-ai/openhands/issues/6729
16:15:55 - openhands:WARNING: issue_definitions.py:597 - Failed to fetch issue 6218: 404 Client Error: Not Found for url: https://api.github.com/repos/all-hands-ai/openhands/issues/6218
16:15:55 - openhands:WARNING: issue_definitions.py:597 - Failed to fetch issue 6668: 404 Client Error: Not Found for url: https://api.github.com/repos/all-hands-ai/openhands/issues/6668
16:15:56 - openhands:WARNING: issue_definitions.py:597 - Failed to fetch issue 6605: 404 Client Error: Not Found for url: https://api.github.com/repos/all-hands-ai/openhands/issues/6605
16:15:56 - openhands:WARNING: issue_definitions.py:597 - Failed to fetch issue 6608: 404 Client Error: Not Found for url: https://api.github.com/repos/all-hands-ai/openhands/issues/6608
16:15:56 - openhands:WARNING: issue_definitions.py:597 - Failed to fetch issue 6615: 404 Client Error: Not Found for url: https://api.github.com/repos/all-hands-ai/openhands/issues/6615
16:15:56 - openhands:WARNING: issue_definitions.py:597 - Failed to fetch issue 6680: 404 Client Error: Not Found for url: https://api.github.com/repos/all-hands-ai/openhands/issues/6680
16:15:56 - openhands:WARNING: issue_definitions.py:597 - Failed to fetch issue 6489: 404 Client Error: Not Found for url: https://api.github.com/repos/all-hands-ai/openhands/issues/6489
16:15:56 - openhands:WARNING: issue_definitions.py:597 - Failed to fetch issue 6617: 404 Client Error: Not Found for url: https://api.github.com/repos/all-hands-ai/openhands/issues/6617
16:15:57 - openhands:WARNING: issue_definitions.py:597 - Failed to fetch issue 6683: 404 Client Error: Not Found for url: https://api.github.com/repos/all-hands-ai/openhands/issues/6683
16:15:57 - openhands:WARNING: issue_definitions.py:597 - Failed to fetch issue 6619: 404 Client Error: Not Found for url: https://api.github.com/repos/all-hands-ai/openhands/issues/6619
16:15:57 - openhands:WARNING: issue_definitions.py:597 - Failed to fetch issue 6633: 404 Client Error: Not Found for url: https://api.github.com/repos/all-hands-ai/openhands/issues/6633
16:15:57 - openhands:WARNING: issue_definitions.py:597 - Failed to fetch issue 6634: 404 Client Error: Not Found for url: https://api.github.com/repos/all-hands-ai/openhands/issues/6634
16:15:58 - openhands:WARNING: issue_definitions.py:597 - Failed to fetch issue 6637: 404 Client Error: Not Found for url: https://api.github.com/repos/all-hands-ai/openhands/issues/6637
16:15:58 - openhands:WARNING: issue_definitions.py:597 - Failed to fetch issue 6640: 404 Client Error: Not Found for url: https://api.github.com/repos/all-hands-ai/openhands/issues/6640
16:15:58 - openhands:WARNING: issue_definitions.py:597 - Failed to fetch issue 6644: 404 Client Error: Not Found for url: https://api.github.com/repos/all-hands-ai/openhands/issues/6644
16:15:59 - openhands:WARNING: issue_definitions.py:597 - Failed to fetch issue 6714: 404 Client Error: Not Found for url: https://api.github.com/repos/all-hands-ai/openhands/issues/6714
16:15:59 - openhands:WARNING: issue_definitions.py:597 - Failed to fetch issue 6590: 404 Client Error: Not Found for url: https://api.github.com/repos/all-hands-ai/openhands/issues/6590
```

We should write a test to reproduce this issue, and fix it.

### OpenHands Installation

Docker command in README

### OpenHands Version

_No response_

### Operating System

None

### Logs, Errors, Screenshots, and Additional Context

_No response_

### PR DESCRIPTION
**Pull request by neubig**
This PR improves the GitHub resolver by:

1. Making API response handling more robust:
   - Handle both list and single-object responses from GitHub API
   - Add proper error handling for comment fetching

2. Optimizing PR handling:
   - Fetch PRs directly by number instead of filtering from all issues
   - Add error handling for individual PR fetches

3. Adding comprehensive test coverage:
   - Test single-object API responses
   - Test error handling in comment fetching
   - Test PR metadata handling
   - Test duplicate issue references
   - Test review thread comments

Note: This PR was originally titled as fixing #5456, but upon review, these changes do not directly address that issue. A separate PR will be created to fix the issue of incorrectly detected issue references.

---

To run this PR locally, use the following command:
```
docker run -it --rm   -p 3000:3000   -v /var/run/docker.sock:/var/run/docker.sock   --add-host host.docker.internal:host-gateway   -e SANDBOX_RUNTIME_CONTAINER_IMAGE=docker.all-hands.dev/all-hands-ai/runtime:c42164b-nikolaik   --name openhands-app-c42164b   docker.all-hands.dev/all-hands-ai/openhands:c42164b
```

### CODE REVIEW

File: openhands/resolver/issue_definitions.py [hunk 1]
```diff
@@ -509,57 +504,69 @@ def get_converted_issues(
         if not issue_numbers:
             raise ValueError('Unspecified issue numbers')
 
-        all_issues = self._download_issues_from_github()
```
Comment (enyst) [inline openhands/resolver/issue_definitions.py:512]: Great catch! 😅

File: openhands/resolver/issue_definitions.py [hunk 2]
```diff
@@ -509,57 +504,69 @@ def get_converted_issues(
         if not issue_numbers:
```
Comment (enyst) [inline openhands/resolver/issue_definitions.py:504]: just a thought, if this must not be None, maybe we can fix the type hint too?

**Review by neubig — COMMENTED** [review]
I think this looks OK to me now

File: openhands/resolver/issue_definitions.py [hunk 3]
```diff
@@ -544,57 +562,86 @@ def get_converted_issues(
...
-            )
+        for issue_number in issue_numbers:
+            try:
+                url = f'https://api.github.com/repos/{self.owner}/{self.repo}/issues/{issue_number}'
+                response = requests.get(url, headers=headers)
+                response.raise_for_status()
+                issue = response.json()
+
+                # Handle both list and single-object responses
+                if isinstance(issue, list):
```
Comment (malhotra5) [inline openhands/resolver/issue_definitions.py:580]: This section of code that checks for list and dict instance seems to be a duplicate in a `IssueHandler`

Ideally we share this function across the handlers

**Review by malhotra5 — CHANGES_REQUESTED** [review]

File: openhands/resolver/issue_definitions.py (hunk 3 above)
Comment (malhotra5) [reply openhands/resolver/issue_definitions.py:580]: @openhands-agent please take a look at this

**Comment by openhands-agent** [comment]
[OpenHands](https://github.com/All-Hands-AI/OpenHands) started fixing the pr! You can monitor the progress [here](https://github.com/All-Hands-AI/OpenHands/actions/runs/12530414654).

**Comment by openhands-agent** [comment]
Status Overview:

❌ Not Resolved: The core issue of code duplication in list/dict instance checking logic between handlers remains unaddressed.

Key Points:
- Original feedback requested consolidation of duplicated validation logic
- Response focused on unrelated issue reference extraction changes
- No shared utility function was created
- Handlers still contain duplicated code

Next Steps Required:
1. Create shared utility function for list/dict validation
2. Refactor handlers to use the shared function
3. Verify functionality remains intact

File: openhands/resolver/issue_definitions.py (hunk 3 above)
Comment (openhands-agent) [reply openhands/resolver/issue_definitions.py:584]: Openhands fix success summary


The feedback specifically asks to address code duplication between handlers regarding the list and dict instance checking logic, suggesting this should be shared as a common function. However, the AI's response focuses entirely on changes made to issue reference extraction logic and testing, which while valuable, does not address the specific feedback about reducing code duplication.

The AI should have:
1. Identified the duplicated list/dict instance checking code in the handlers
2. Created a shared utility function that both handlers could use
3. Refactored the handlers to use this shared function
4. Verified the changes maintain the same functionality

Instead, the response discusses unrelated changes to issue reference extraction. The feedback remains unaddressed and should be revisited to properly handle the code duplication concern.

**Comment by neubig** [comment]
I'm actually not sure if this is a problem before, maybe we can close this but I'd be happy to re-open if necessary.

**Pull request closed without merging**